### PR TITLE
Added container_labels parameter

### LINKED ISSUE #1263: Docker plugin together with Prometheus fails for some containers due to inconsistent label cardinality
### System info:
Amazon Linux AMI release 2016.03 (based on RHEL)
Docker: Version: 1.9.1 (build a34a1d5/1.9.1), API version:  1.21
Telegraf - version 0.13.0

### Steps to reproduce:

1. Run Telegraf with docker plugin, prometheus output, and containers which do not have the same labels.

### Expected behavior:

Prometheus should report stats about all containers.

### Actual behavior:

Prometheus output reports only stats about some containers.

### Additional info:

The docker plugin reports only part of my containers, because for most it reports  inconsistent label cardinality, e.g.:

`2016/05/24 18:58:30 ERROR Getting metric in Prometheus output, key: docker_container_net_tx_errors, labels: map[network:eth0 container_image:SOMEIMAGE container_name:SOMENAME host:SOMEIP],
err: inconsistent label cardinality`

When I look at containers that are successfully reported in prometheus output:
`docker_container_net_tx_packets{License="GPLv2",Vendor="CentOS",container_image="SOMEIMAGE2",container_name="SOMENAME2",host="SOMEIP",network="eth0"} 773869`
I am seeing that in this case the container has additional labels License and Vendor. 

Verifying with `curl --unix-socket /var/run/docker.sock http:/containers/json | jq '.[].Labels'` they indeed have inconsistent labels:
`{}
{}
{}
{
  "build-date": "2016-03-31",
  "license": "GPLv2",
  "name": "CentOS Base Image",
  "vendor": "CentOS"
}
{
  "License": "GPLv2",
  "Vendor": "CentOS"
}
{
  "License": "GPLv2",
  "Vendor": "CentOS"
}
{
  "License": "GPLv2",
  "Vendor": "CentOS"
}`

Docker documentation tells that arbitrary labels can be set during build and container creation. A workaround would be to make sure all my containers have the same label set, but unfortunately labels can not be changed/removed for a running containers. 

## Proposal:

Quick: Do not add the container labels reported for containers by docker as metric labels. I assume that like in this case these are pretty much irrelevant in most cases. 

Bonus: Make the labels the docker plugin grabs from the container labels configurable and make sure that if a label is not present it still gets added with a default (e.g. "missing") value. Thus, if someone wants to use labels for some fancy grouping, one could do that.

### PR DESCRIPTION
**Pull request by noxer**
Fixes #1263

### CODE REVIEW

File: plugins/inputs/docker/docker.go [hunk 1]
```diff
@@ -56,6 +57,9 @@ var sampleConfig = `
   endpoint = "unix:///var/run/docker.sock"
   ## Only collect metrics for these containers, collect all if empty
   container_names = []
+  ## Only collect these container labels from docker daemon, collect all if empty but note that
```
Comment (sparrc) [inline plugins/inputs/docker/docker.go]: please remove the mention of the prometheus output, this is too specific for an option that is in fact general.

**Comment by sparrc** [comment]
you should have the container_labels array support glob matching using github.com/gobwas/glob.

If the user specifies multiple patterns in container_labels you can construct a list of matches similar to how we do filtering here: https://github.com/influxdata/telegraf/blob/master/internal/models/filter.go#L87

you should only compile the filter once.

**Comment by noxer** [comment]
How would you implement inserting placeholders with those filters?

**Comment by sparrc** [comment]
@noxer what do you mean by that?

**Comment by noxer** [comment]
I'm inserting placeholders ("-") for labels that are in d.ContainerLabels but not in container.Labels: `tags[l] = "-"`. Using glob would make that impossible (I can't insert something for "*xyz"). I would suggest inserting such filters as a separate parameter.

**Comment by sparrc** [comment]
hmm I see what you mean. I'd actually rather not merge this, those placeholder labels will appear in every output plugin but don't mean anything except in the prometheus case. I think that you need to come up with a solution that is specific to the prometheus output.

**Comment by noxer** [comment]
@sparrc This seems to be a problem specific to docker input + prometheus output, prometheus needs a fixed set of tags (per input plugin) and docker generates a variable set of tags (due to the container labels). At least in our setup docker is the only input plugin that does that.

**Comment by sparrc** [comment]
this is true for your particular use-case, but in reality it could be any of the more than 70 input plugins that generate variable numbers of tags.

**Comment by lightblu** [comment]
These placeholders will only appear for nonexistent labels which where specified in container_labels, if you do not specify container_labels you get completely the current behavior and no placeholders at all.

I am still not sure if it is generally ok for a plugin to provide inconsistent and arbitrary labels for a plugin/metric, because no other input plugin I used so far does this (mongo, rabbitmq, and all the default ones like mem, cpu, disk etc.).

I just found that what you propose may be possible with Telegrafs configuration (taginclude) alone, e.g. specifying taginclude = ['network', and all other ones that are not extracted from the container labels, because I need those] on the docker plugin, still it feels wrong for me that you have to specify this to not break an output plugin and to allow for inconsistent and arbitrary labels.  Will try this now.

**Comment by noxer** [comment]
Turns out taginclude does exactly what we need, making this PR obsolete.

**Comment by sparrc** [comment]
@lightblu the labels are not "arbitrary", _you_ are specifying them as docker labels 😑

as I said before, this is only a problem for prometheus. It doesn't make sense and it's very bad programming design to have 100 little workarounds in every single input plugin just for the sake of a single output plugin.

**Comment by lightblu** [comment]
I agree its fine if this is the general approach within Telegraf/Influx, but to me it looked like bad design to have inconsistent label sets on the same metric (biased by prometheus) and the docker plugin to be the only input plugin which shows this behaviour (although I am wrong with this it seems), if this is the general approach I'm sorry.
But the labels are arbitrary in the sense that you usually do not build up all your containers from scratch but use prebuilt ones from registries, where you are bound to the arbitrary labels the creators have already added (see examples in the issue), you cannot remove these when running containers but just add/overwrite.

**Pull request closed without merging**